rust/kernel/chrdev.rs

```diff
@@ -20,10 +20,62 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+struct SafeCdev(*mut bindings::cdev);
+
+impl SafeCdev {
+    fn alloc() -> KernelResult<Self> {
+        // SAFETY: call an unsafe function
+        let cdev = unsafe { bindings::cdev_alloc() };
+        if cdev.is_null() {
+            return Err(Error::ENOMEM);
+        }
+        Ok(Self(cdev))
+    }
+
+    fn init(&mut self, fops: *const bindings::file_operations) {
+        // SAFETY: call an unsafe function
+        unsafe {
+            bindings::cdev_init(self.0, fops);
+        }
+    }
+
+    fn add(&mut self, dev: bindings::dev_t, pos: c_types::c_uint) -> c_types::c_int {
+        // SAFETY: call an unsafe function
+        unsafe { bindings::cdev_add(self.0, dev, pos) }
+    }
+
+    fn set_owner(&mut self, module: &crate::ThisModule) {
+        // SAFETY: dereference of a raw pointer
+        unsafe {
+            (*self.0).owner = module.0;
+        }
+    }
+}
+
+fn new_none_array<T, const N: usize>() -> [Option<T>; N] {
+    // SAFETY: manipulate MaybeUninit memory
+    unsafe {
+        let mut arr: [MaybeUninit<Option<T>>; N] = MaybeUninit::uninit_array();
+        for elem in &mut arr {
+            elem.as_mut_ptr().write(None);
+        }
+        MaybeUninit::array_assume_init(arr)
+    }
+}
```

Comment:
This function looks pretty bad. I couldn't work out how to initialize an `[Option<SafeCdev>; N]` array to `None`:
```rust
cdevs: [None; N],
```
Rust insists that `SafeCdev` needs the `Copy` trait, which is obviously a really bad idea for a refcounted object:
```
error[E0277]: the trait bound `Option<SafeCdev>: core::marker::Copy` is not satisfied
   --> rust/kernel/chrdev.rs:155:24
    |
155 |                 cdevs: [None; N],
    |                        ^^^^^^^^^ the trait `core::marker::Copy` is not implemented for `Option<SafeCdev>`
    |
    = help: the following implementations were found:
              <Option<T> as core::marker::Copy>
    = note: the `Copy` trait is required because the repeated element will be copied
```

Does anyone know of an elegant, safe way to accomplish this?